cmail-admin/doc/cmail-admin-address.8
.TH CMAIL-ADMIN-ADDRESS 8 "August 2015" "v0.1" "cmail Administration Tools"
.SH NAME
cmail-admin-address \- Configure inbound path expressions for cmail
.SH SYNOPSIS
.BI "cmail-admin-address [ --dbpath " "master-database-file" " ] [ --verbosity " N
.BI " | -v " N " ] [ --help | -h ] " "mode" " "  "mode-options"
.SH DESCRIPTION
.B cmail-admin-address
allows the administrator of a cmail server system to create, modify and delete
address entries, which are tuples of an address expression, an address order
and routing information describing how mail adressed to the path is to be handled. 
Inbound paths are matched against the set of all configured entries, ordered totally by the 
address order, with the highest order value being matched first.
Expressions may contain arbitrary-length wildcards (
.B %
) and single-character wildcards (
.B _
). Further information about the exact matching semantics may be found by referring
to the 
.B SEE ALSO 
section.

.SH OPTIONS
.TP 
.BI "--dbpath " master-database-file ", -d " master-database-file
Select an alternate master database file instead of the default
.IR /etc/cmail/databases/master.db3 .
This overrides the database path provided by the
.B CMAIL_MASTER_DB
environment variable.
.TP
.BI "--verbose " N ", -v " N
Increase verbosity on a scale of
.B 0
to 
.B 4
.TP
.B "--help, -h"
Print helpful hints about operating the program
.SH FILES
.I /etc/cmail/databases/master.db3
.RS
Default path of the master database when installed system-wide. May be overridden either by specifying the
.B CMAIL_MASTER_DB
environment variable or by supplying the
.BR "--dbpath " parameter.
.SH MODES

.TP
.BI "list " [expression]
List all address entries contained in the database. Optionally, filter the list for entries matching
.IR expression "."

.TP
.BI "test " mailpath
Test which address expressions would match an incoming 
.IR mailpath "."
.RS
.B Example:
$ cmail-admin-address test cb@cbcdn.com
.RE

.TP
.BI "add " "expression router [router-argument [order]]"
Add an inbound path entry for the provided
.IR "expression" ", to be handled by the specified " router ". Some routers may require a " router-argument "."
.RI "Optionally, an " order " value may be provided, which will be assigned to the path unless there is already an active entry using that " order
.RS
.B Example:
$ cmail-admin-address add "%@cbcdn.com" store cbdev
.RE

.TP
.BI "delete " "order"
Delete the expression entry assigned the given
.IR order .
.RS
.B Example:
$ cmail-admin-address delete 9001
.RE

.TP
.BI "swap " "first second"
.RB "Swap the " order " values of the two entries indicated by the given order values " 
.IR first " and " second "."
.RS
.B Example:
$ cmail-admin-address swap 9001 9002
.RE

.SH ROUTERS
Inbound routers are processing function applied to mail for each path in the recipients list, some of wich take an additional argument.

.TP
.BI "store " username
.RI "Store the received mail to the inbox of " username ". Note that this does not resolve aliases. If "
.IR username " is not set or not a valid user, this router behaves the same way as the "
.BR reject " router."

.TP
.BI "handoff " remote-spec
Have incoming mail for this path be handled by a remote mail server
.BR "(smart host) " "by simply reusing the original envelope with the new remote. If "
.IR remote-spec " is not set, this router behaves the same way as the
.BR reject " router. If the "
.IR remote-spec " is invalid, the outbound smtp client will generate a bounce notification message after "
a configured amount of retries.

.TP
.BI "redirect " new-recipient
.RI "Rewrite the envelope recipient of the mail according to " new-recipient ", which may optionally contain"
any of the following variables:
.RS
.BR "	(to-local)" "	will be replaced with the local part of the recipient path"
.RE
.RS
.BR "	(from-local)" "	will be replaced with the local part of the reverse path"
.RE
.RS
.BR "	(to-domain)" "	will be replaced with the domain part of the recipient path"
.RE
.RS
.BR "	(from-domain)" "	will be replaced with the domain part of the reverse path"
.RE

.RS
.RI "If " new-recipient " is not set, this router will behave the same way as the"
.BR reject " router. If " new-recipient " is an invalid path, the outbound smtp client will generate"
a bounce notification message after a configured amount of retries.

Note that no router rewrites the header or body section of a mail. Therefore, the original
recipient may still appear in the relevant headers.
.RE

.TP
.BI "reject " [reason]
Reject the recipient path (SMTP 551), optionally presenting
.IR reason " as failure reason."

.TP
.B drop
Accept but quietly drop incoming mail for this recipient path.

.SH BUGS
For a discussion of pitfalls and possible sources of misconfigurations refer to the CAVEATS.txt file in the source distribution.
.SH AUTHORS
 Jan "mpease" Duepmeier <me at mpease dot de>  
 Fabian "cbdev" Stumpf <fjs at fabianstumpf dot de>  
.SH "SEE ALSO"
Online documentation and homepage at http://cmail.rocks/
.PP
SQLite "LIKE" syntax: https://www.sqlite.org/lang_expr.html#like
.PP
.BR cmail-admin-user (8),
.BR cmail-admin-smtpd (8),
.BR cmail-admin-popd (8),
.BR cmail-smtpd (8),
.BR cmail-popd (8),
.BR cmail-dispatchd (8),
.BR cmail-master-database (5)



debian/changelog
cmail-mta (1.0) unstable; urgency=low

  * Initial Release.

 -- root <root@unknown>  Sat, 18 Jul 2015 14:24:27 +0000


cmail-admin/address/cmail-admin-address.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlite3.h>
#include <time.h>

#include "../cmail-admin.h"
#include "address.c"

#define PROGRAM_NAME "cmail-admin-address"

int usage(char* fn) {
	printf("cmail-admin-address - Configure inbound address routing\n");
	printf("This program is part of the cmail administration toolkit\n");
	printf("Usage: %s <options> <commands> <arguments>\n\n", fn);

	printf("Basic options:\n");
	printf("\t--verbosity, -v\t\tSet verbosity level (0 - 4)\n");
	printf("\t--dbpath, -d <dbpath>\tSet master database path (Default: %s)\n", DEFAULT_DBPATH);
	printf("\t--help, -h\t\tDisplay this help message\n");

	printf("Commands:\n");
	printf("\tadd <expression> <router> [<router argument> [<order>]]\t\tCreate an inbound path\n");
	printf("\tdelete <order>\t\t\t\t\t\t\tDelete a path expression\n");
	printf("\tupdate <order> <router> [<route argument>]\t\t\tUpdate routing information for a path\n");
	printf("\tswap <first> <second>\t\t\t\t\t\tSwap order of two expressions (given by their orders)\n");
	printf("\ttest <mailpath>\t\t\t\t\t\tTest <mailpath> against the address expression database\n");
	printf("\tlist [<expression>]\t\t\t\t\t\tList all currently configured paths, optionally filtered by <expression>\n");
	return 1;
}

#include "../lib/common.c"


int mode_add(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	int status = 0;

	if (argc < 3) {
		logprintf(log, LOG_ERROR, "Missing arguments\n\n");
		return -1;
	}

	if(!router_valid(argv[2])){
		logprintf(log, LOG_WARNING, "Unknown router %s, adding anyway\n", argv[2]);
	}

	if (argc == 3) {
		status = sqlite_add_address(log, db, argv[1], argv[2], NULL);
	} else if (argc == 4) {
		status = sqlite_add_address(log, db, argv[1], argv[2], argv[3]);
	}
	else if (argc == 5) {
		status = sqlite_add_address_order(log, db, argv[1], strtol(argv[4], NULL, 10), argv[2], argv[3]);
	}
	else{
		logprintf(log, LOG_ERROR, "Too many arguments\n\n");
		status = -1;
	}

	return status;
}

int mode_delete(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	if (argc < 2) {
		logprintf(log, LOG_ERROR, "Missing argument\n\n");
		return -1;
	}

	return sqlite_delete_address(log, db, strtol(argv[1], NULL, 10));
}

int mode_swap(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	if (argc < 3) {
		logprintf(log, LOG_ERROR, "Missing arguments\n\n");
		return -1;
	}

	return sqlite_swap(log, db, strtol(argv[1], NULL, 10), strtol(argv[2], NULL, 10));
}

int mode_update(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	if (argc < 3) {
		logprintf(log, LOG_ERROR, "Missing arguments\n\n");
		return -1;
	}

	if(!router_valid(argv[2])){
		logprintf(log, LOG_WARNING, "Unknown router %s, adding anyway\n", argv[2]);
	}

	if(argc == 3){
		return sqlite_update_address(log, db, strtol(argv[1], NULL, 10), argv[2], NULL);
	}
	else{
		return sqlite_update_address(log, db, strtol(argv[1], NULL, 10), argv[2], argv[3]);
	}
}

int mode_list(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	char* filter = "%";

	if (argc > 1) {
		filter = argv[1];
	}

	return sqlite_get_address(log, db, filter, false);
}

int mode_test(LOGGER log, sqlite3* db, int argc, char* argv[]) {
	char* filter = "%";

	if (argc > 1) {
		filter = argv[1];
	}

	return sqlite_get_address(log, db, filter, true);
}

int main(int argc, char* argv[]) {
	struct config config = {
		.verbosity = 0,
		.dbpath = getenv("CMAIL_MASTER_DB")
	};


	if (!config.dbpath) {
		config.dbpath = DEFAULT_DBPATH;
	}

	// argument parsing
	add_args();

	char* cmds[argc];
	int cmdsc = eargs_parse(argc, argv, cmds, &config);

	LOGGER log = {
		.stream = stderr,
		.verbosity = config.verbosity
	};

	if (cmdsc < 0) {
		return 1;
	}

	sqlite3* db = NULL;

	if(!cmdsc){
		logprintf(log, LOG_ERROR, "No command specified\n\n");
		exit(usage(argv[0]));
	}

	logprintf(log, LOG_INFO, "Opening database at %s\n", config.dbpath);
	db = database_open(log, config.dbpath, SQLITE_OPEN_READWRITE);

	if(!db){
		exit(usage(argv[0]));
	}

	//check database version
	if (database_schema_version(log, db) != CMAIL_CURRENT_SCHEMA_VERSION) {
		logprintf(log, LOG_ERROR, "The specified database (%s) is at an unsupported schema version.\n");
		sqlite3_close(db);
		return 11;
	}

	int status;

	if (!strcmp(cmds[0], "add")) {
		status = mode_add(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "delete")) {
		status = mode_delete(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "swap")) {
		status = mode_swap(log, db, cmdsc, cmds);
	} else if (!strcmp(cmds[0], "update")) {
		status = mode_update(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "list")) {
		status = mode_list(log, db, cmdsc, cmds);
	}  else if (!strcmp(cmds[0], "test")) {
		status = mode_test(log, db, cmdsc, cmds);
	} else {
		logprintf(log, LOG_ERROR, "Unkown command %s\n\n", cmds[0]);
		exit(usage(argv[0]));
	}

	sqlite3_close(db);

	return status;
}
